src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +1010,247 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+        {
+            double startPoint;
+            pToDeck->fadeBeginPos = getLastSoundPosition(pToDeck);
+            if (seekToStartPoint || toDeckPosition >= pToDeck->fadeBeginPos) {
+                    // toDeckPosition >= pToDeck->fadeBeginPos happens when the
+                    // user has seeked or played the to track behind fadeBeginPos of
+                    // the fade after the next.
+                    // In this case we recue the track just before the transition.
+                startPoint = getFirstSoundPosition(pToDeck);
+            } else {
+                startPoint = toDeckPosition;
+            }
+            useFixedFadeTime(pFromDeck, pToDeck,
+                    getLastSoundPosition(pFromDeck),
+                    startPoint);
+        }
+        break;
+    case TransitionMode::FixedFullTrack:
+    default:
+        {
+            double startPoint;
+            pToDeck->fadeBeginPos = pToDeck->duration();
+            if (seekToStartPoint || toDeckPosition >= pToDeck->fadeBeginPos) {
+                    // toDeckPosition >= pToDeck->fadeBeginPos happens when the
+                    // user has seeked or played the to track behind fadeBeginPos of
+                    // the fade after the next.
+                    // In this case we recue the track just before the transition.
+                startPoint = 0;
+            } else {
+                startPoint = toDeckPosition;
+            }
+            useFixedFadeTime(pFromDeck, pToDeck,
+                    pFromDeck->duration(),
+                    startPoint);
+        }
+    }
 
-        if (toTrackDuration > 0.0) {
-            // Transition time is negative here, so we need to add instead of subtract.
-            pToDeck->startPos = (toTrackIntroStart + m_transitionTime) / toTrackDuration;
-        } else {
-            pToDeck->startPos = 0.0;
+    // These are expected to be a fraction of the track length.
+    pFromDeck->fadeBeginPos /= fromTrackDuration;
+    pFromDeck->fadeEndPos /= fromTrackDuration;
+    pToDeck->startPos /= toTrackDuration;
+    pToDeck->fadeBeginPos /= toTrackDuration;
+
+    pFromDeck->isFromDeck = true;
+    pToDeck->isFromDeck = false;
+
+    DEBUG_ASSERT(pFromDeck->fadeBeginPos <= 1);
+
+    if (sDebug) {
+        qDebug() << this << "calculateTransition" << pFromDeck->group
+                << pFromDeck->fadeBeginPos << pFromDeck->fadeEndPos
+                << pToDeck->startPos;
+    }
+}
+
+void AutoDJProcessor::useFixedFadeTime(DeckAttributes* pFromDeck,
+        DeckAttributes* pToDeck, double endPoint, double startPoint) {
+    if (m_transitionTime > 0.0) {
+        // Guard against the next track being too short. This transition must finish
+        // before the next transition starts.
+        double toDeckOutroStart = getOutroStartPosition(pToDeck);
+        if (toDeckOutroStart <= startPoint) {
+            // we are already too late
+            double end = getOutroEndPosition(pToDeck);
```

Comment:
This should become a parameter to use the correct value for the selected mode.

Reply:
I am not entirely sure how you want this to work, so please jump in to fix it.